### PR TITLE
[WIP]: add detection of physical cores and ability to limit jobs to the number of physical (rather than logical) cores

### LINKED ISSUE #2542: vboxwrapper should use ncpus = physical cores
From https://www.virtualbox.org/manual/ch03.html:

> You should not, however, configure virtual machines to use more CPU cores than you have available physically (real cores, no hyperthreads).

However currently vboxwrapper uses the number logical cores. There is some discussion of this in https://github.com/BOINC/boinc/issues/1501 and https://github.com/marius311/boinc-server-docker/issues/33.

I had thought this was some minor sub-optimality, but thanks to @zonca getting a totally reproducible scenario working, I recently made the connection that this could be causing a significant number of the failures we see at C@H which I had previously not understood. 

I don't have a solution, see https://github.com/BOINC/boinc/issues/1501 for discussion of the issues involved in getting the number of physical cores (anyone know a good API?). I'm opening this though so we can brainstorm / track progress on this specific issue.

### PR DESCRIPTION
**Pull request by marius311**
WIP, do not merge. 

- [ ] Client detection
  - [x] Linux
  - [ ] Windows
  - [ ] Mac
- [x] Receive info on server, update DB, and display in host info pages
- [ ] Add planclass (?) to allow limiting to physical cores
- [ ] Use this info in scheduling

Fixes #2542

### CODE REVIEW

File: client/hostinfo_unix.cpp [hunk 1]
```diff
@@ -39,6 +39,8 @@
 #include <vector>
 #include <string>
 #include <cstring>
+#include <tuple>
```
Comment (AenBleidd) [inline client/hostinfo_unix.cpp]: I'm not sure this is available in VS2010

Comment (marius311) [reply client/hostinfo_unix.cpp]: Thanks, didn't know. That's annoying. So the client needs to be buildable with VS2010 on Windows? What's the motivation for supporting such an old version?

Comment (marius311) [reply client/hostinfo_unix.cpp]: Nevermind, I've got the VS2010 story explained to me at the workshop :) For now I'll get a workaround.

Comment (ChristianBeer) [reply client/hostinfo_unix.cpp]: `std::tuple` is also a C++11 extension and thus not available for the Linux part too.

**Comment by davidpanderson** [comment]
For a plan class with <physical_cpus>, a job might get --nthreads 4 but use 8 logical CPUs (client scheduling is in terms of logical CPUs).  The volunteer would see that the job uses 8 CPUs, but the CPU meter would show 50% usage.  This may be confusing to overly attentive volunteers.

**Comment by davidpanderson** [comment]
The semantics of <max_threads>N 1</max_threads> is that the job uses the
number of physical CPUs minus one, i.e. leave one free to avoid slowing down desktop response.
If the user has specified "use X% of the CPUs" in their computing prefs with X < 100,
this is probably not the desired behavior; there will already be a free CPU.
I propose changing the semantics so that the minus one only applies to hosts
where all CPUs are usable.

**Comment by marius311** [comment]
Sorry I've not had time to finish this up. I will try this week. 

> I propose changing the semantics so that the minus one only applies to hosts
where all CPUs are usable.

Seems reasonable to me. 

> For a plan class with <physical_cpus>, a job might get --nthreads 4 but use 8 logical CPUs (client scheduling is in terms of logical CPUs). The volunteer would see that the job uses 8 CPUs, but the CPU meter would show 50% usage. This may be confusing to overly attentive volunteers.

Can you explain in more detail why this would happen? I don't quite follow, and this seems a pretty serious deficiency.

**Comment by davidpanderson** [comment]
The Windows Task Manager (the default CPU monitor) shows logical CPU usage.
So if an app is using 4 physical cores, the task manager will show 50% CPU usage.

This is not a big deal, and there's nothing we can do about it except to explain
to users what's going on.

**Comment by marius311** [comment]
> The Windows Task Manager (the default CPU monitor) shows logical CPU usage.
So if an app is using 4 physical cores, the task manager will show 50% CPU usage.

OK, I'm on the same page with that. 

However, the way I was thinking about this is that on a 4 phys core / 8 logical core machine, if the user's CPU preference was to 100%, then they could run e.g.

* two separate nthreads=4 vboxwrapper jobs, or 
* one nthreads=4 vboxwrapper job, and the remaining 4 logical cores could run other BOINC jobs if that's what happened to be available

In either case, the vboxwrapper job should show "running with 4 CPUs" in the boinc manager. 

Basically, I don't think that `<physical_cpus>` should imply that the job needs exclusive use of the entire physical CPU, because, at least as the Virtualbox docs are written, this is not a requirement of Virtualbox. However, I realize now we've never actually tested this. I've asked a colleague to test this here https://github.com/marius311/boinc-server-docker/issues/33#issuecomment-410889201, his reply could inform what we do.

**Comment by davidpanderson** [comment]
The way I implemented things, the client will run only one 4-thread VM;
i.e. the client will treat it like it's using 8 logical CPUs (and will show
it as such).

Changing job scheduling on the client to distinguish between logical and
physical CPUs
would be a large job, which I suggest we defer.

**Comment by davidpanderson** [comment]
I'm not sure the <max_threads>N M<max_threads> semantics are useful after all.
You can tell it to not use a CPU (M=1).
But the client will then run a single-thread job on that CPU.
The only way to keep CPUs free is with the "use at most X% of the CPUs" preference.

**Comment by marius311** [comment]
> But the client will then run a single-thread job on that CPU.

No that's OK, as long as that ends up not being a VBox job. The problem is that the VBox job can't be set to low priority, but a normal native job filling that spot can be, so the user would suffer from less desktop degradation. Lots of C@H users currently have this set up succesfully. 

> The way I implemented things, the client will run only one 4-thread VM;

Yea, I see that now. I admit I don't totally understand scheduling, but why can't the new system simply set, for these vbox jobs, nthreads_max=4, and then let the client pack two of them in like I think would happen with any other two native nthreads=4 jobs ? I worry many users are going to think we're wasting 50% of their resources and hate VBox even more than they already do.

**Comment by davidpanderson** [comment]
We could.

Running 2 VMs at once would have twice the working set size,
which could cause increased paging.
I'm curious to know whether (on a typical Win PC)
- 2 VMs gives more or less throughput
- 2 VMs slows down interactive response.
Is there any way to set up such a test?

**Comment by davidpanderson** [comment]
This question applies to FP-intensive apps in general, on hyperthreaded
machines:
does using all the virtual CPUs give more or less throughput?

**Comment by ChristianBeer** [comment]
From time to time there is some intensive testing in the German Rechenkraft.net forum about the effect of hyperthreading on different CPU and different projects. For VirtualBox projects @JacobWKlein did a good comparison and found out that disabling Hyperthreading helps the VBox jobs to complete faster. This is also the case for other science applications (but not all).

**Comment by RichardHaselgrove** [comment]
The question is frequently discussed at SETI, but without any methodical testing. The general consensus is that with HT active, ordinary CPU tasks run slower, but total throughput is greater because more tasks run concurrently. My personal view is that we ignore the differences between integer-heavy and FP-heavy applications at our peril - especially on hardware with FPU shared between logical cores.

**Comment by marius311** [comment]
I would suggest we decouple,

1) The ability to limit certain jobs to run with a per-job max nthreads = (# of phys cores). This is needed by VBox jobs so they don't outright crash. 

2) The ability for the user or for project admins to (potentially) improve performance for VBox or native jobs by running without hyperthreading. 

I think (1) would should be the goal of this PR, then we can potentially think about (2) (of course we can keep it in mind while coming up with the way to do (1)) @ChristianBeer and @RichardHaselgrove your guys' info is helpful suggesting we could indeed want ((2)). 

So if you agree, then I suggest @davidpanderson change the scheduling to just limit <phys_core> jobs to (# of phys cores), but otherwise allow the client to pack multiple of them together.

**Comment by lfield** [comment]
What is that state of this? There hasn't been any update for a while.

**Comment by marius311** [comment]
If I remember correctly, in writing this I used some C++11 extensions which I didn't know I shouldn't. But it was explained to me that someone was working on updating the compilation process, I think using a newer version of Visual Studio or something like that, so that this would be fine. I don't know what's the status of this work. Does anyone know or can link and Issue/PR?

**Comment by AenBleidd** [comment]
@marius311, #2348
But I'm not sure that there are some changes or finished implementations, @JuhaSointusalo car tell more details.
But meanwhile I'd suggest change the code to be compatible with C++ 2003 to move this functionality further

**Comment by RichardHaselgrove** [comment]
Remember that the current release builds were done with VS2010.

**Comment by CharlieFenton** [comment]
Remember that hostinfo_unix.cpp is also used by Mac builds.

**Comment by marius311** [comment]
> But meanwhile I'd suggest change the code to be compatible with C++ 2003 to move this functionality further

I admit I'm pretty unmotivated to do that. I may try and see if there's anything I can help with in #2348 however, that seems like the better solution anyway.

**Comment by lfield** [comment]
I am closing this PR as more work is needed until it is ready. Discussion can continue on the [issue](https://github.com/BOINC/boinc/issues/2542) while branch [phys_cores](https://github.com/BOINC/boinc/tree/phys_cores) is being worked on.
